### PR TITLE
IGNITE-21159 Deserialization of Instant field in a Marshallable fails

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-21159

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/network/src/main/java/org/apache/ignite/internal/network/serialization/ClassDescriptorFactory.java [hunk 1]
```diff
@@ -244,7 +244,7 @@ private boolean hasReadResolve(Class<? extends Serializable> clazz) {
     }
 
     private boolean hasWriteReplace(Class<? extends Serializable> clazz) {
```
Comment (sashapolo) [inline modules/network/src/main/java/org/apache/ignite/internal/network/serialization/ClassDescriptorFactory.java:246]: Can we inline this method?

File: modules/network/src/main/java/org/apache/ignite/internal/network/serialization/DeclaredType.java [hunk 2]
```diff
@@ -29,11 +29,10 @@ public interface DeclaredType {
     int typeDescriptorId();
 
     /**
-     * Returns {@code true} if the type slot can only host (at runtime) instances of its declared type (and not subtypes),
-     * so the runtime marshalling type is known upfront. This is also true for enums, even though technically their values might
+     * Returns {@code true} if the contents of the type slot can only hove (at runtime) instances serialized as its declared type
```
Comment (sashapolo) [inline modules/network/src/main/java/org/apache/ignite/internal/network/serialization/DeclaredType.java:32]: What's a "type slot"?

Comment (sashapolo) [inline modules/network/src/main/java/org/apache/ignite/internal/network/serialization/DeclaredType.java:32]:
```suggestion
     * Returns {@code true} if the contents of the type slot can only have (at runtime) instances serialized as its declared type
```

File: modules/network/src/test/java/org/apache/ignite/internal/network/serialization/marshal/TestUnmarshalUtils.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import static org.hamcrest.Matchers.notNullValue;
+
+import org.apache.ignite.internal.network.serialization.ClassDescriptor;
+import org.apache.ignite.internal.network.serialization.ClassDescriptorRegistry;
+import org.apache.ignite.internal.network.serialization.DescriptorRegistry;
+
+/**
+ * Test utilities for unmarshalling.
+ */
+class TestUnmarshalUtils {
```
Comment (sashapolo) [inline modules/network/src/test/java/org/apache/ignite/internal/network/serialization/marshal/TestUnmarshalUtils.java:31]: I would prefer this class not to be a utility class, something like `NonNullUnmarshaller`

File: modules/network/src/test/java/org/apache/ignite/network/DefaultMessagingServiceTest.java [hunk 4]
```diff
@@ -343,6 +348,31 @@ private static int extractChannelIdFromThreadName(Thread thread) {
         return Integer.parseInt(matcher.group(1));
     }
 
+    @Test
+    @Timeout(10)
```
Comment (sashapolo) [inline modules/network/src/test/java/org/apache/ignite/network/DefaultMessagingServiceTest.java:352]: Why do you need this timeout? Shall we use a timeout provided by `willBe`?

File: modules/network/src/main/java/org/apache/ignite/internal/network/serialization/DeclaredType.java (hunk 2 above)
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/serialization/DeclaredType.java:32]: It's either a field or a component of an array. So it's a slot for which some 'declared type' is statically known.

File: modules/network/src/test/java/org/apache/ignite/internal/network/serialization/marshal/TestUnmarshalUtils.java (hunk 3 above)
Comment (rpuch) [reply modules/network/src/test/java/org/apache/ignite/internal/network/serialization/marshal/TestUnmarshalUtils.java]: Renamed the class. WDYT?

File: modules/network/src/main/java/org/apache/ignite/internal/network/serialization/DeclaredType.java (hunk 2 above)
Comment (sashapolo) [reply modules/network/src/main/java/org/apache/ignite/internal/network/serialization/DeclaredType.java]: Is this terminology used anywhere else? It is quite confusing...

File: modules/network/src/test/java/org/apache/ignite/internal/network/serialization/marshal/TestUnmarshaling.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+
+import org.apache.ignite.internal.network.serialization.ClassDescriptor;
+import org.apache.ignite.internal.network.serialization.ClassDescriptorRegistry;
+import org.apache.ignite.internal.network.serialization.DescriptorRegistry;
+
+/**
+ * Unmarshalling we do it tests.
```
Comment (sashapolo) [inline modules/network/src/test/java/org/apache/ignite/internal/network/serialization/marshal/TestUnmarshaling.java:29]: This javadoc looks wrong

File: modules/network/src/test/java/org/apache/ignite/internal/network/serialization/marshal/TestUnmarshaling.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+ */
+class TestUnmarshaling {
```
Comment (sashapolo) [inline modules/network/src/test/java/org/apache/ignite/internal/network/serialization/marshal/TestUnmarshaling.java:31]:
```suggestion
class TestUnmarshalling {
```

File: modules/network/src/test/java/org/apache/ignite/internal/network/serialization/marshal/CleanSlateUnmarshaller.java [hunk 7]
```diff
@@ -26,20 +26,27 @@
...
      * @param marshalled Marshalling result.
-     * @param marshaller Marshaller to use.
-     * @param marshallingRegistry Registry that was used when marshalling.
      * @param <T> Expected type of the unmarshalled value.
      * @return Unmarshalled value.
      * @throws UnmarshalException If something goes wrong.
      */
-    static <T> T unmarshalNonNull(MarshalledObject marshalled, UserObjectMarshaller marshaller, DescriptorRegistry marshallingRegistry)
+    <T> T unmarshalNonNull(MarshalledObject marshalled)
             throws UnmarshalException {
```
Comment (sashapolo) [inline modules/network/src/test/java/org/apache/ignite/internal/network/serialization/marshal/CleanSlateUnmarshaller.java:50]: This formatting is a little bit off

**Review by sashapolo — APPROVED** [review]

File: modules/network/src/main/java/org/apache/ignite/internal/network/serialization/DeclaredType.java (hunk 2 above)
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/serialization/DeclaredType.java]: I don't know, I had to invent something myself. Any ideas are welcome :)

**Pull request merged**